dbms/src/DataStreams/CastTypeBlockInputStream.cpp

```diff
@@ -100,20 +100,26 @@ Block CastEnumBlockInputStream::readImpl()
     return res;
 }
 
-void CastEnumBlockInputStream::collectEnums(const Block & in_sample, const Block & out_sample)
+void CastTypeBlockInputStream::collectDifferent(const Block & in_sample, const Block & out_sample)
 {
     size_t in_size = in_sample.columns();
-    enum_types.resize(in_size);
+    cast_types.resize(in_size);
     for (size_t i = 0; i < in_size; ++i)
     {
         const auto & in_elem  = in_sample.getByPosition(i);
         const auto & out_elem = out_sample.getByPosition(i);
 
-        /// Force conversion only if source type is not Enum.
+        /// Force conversion if source type is not Enum.
         if ( dynamic_cast<IDataTypeEnum*>(out_elem.type.get()) &&
             !dynamic_cast<IDataTypeEnum*>(in_elem.type.get()))
         {
-            enum_types[i] = NameAndTypePair(out_elem.name, out_elem.type);
+            cast_types[i] = NameAndTypePair(out_elem.name, out_elem.type);
+        }
+
+        /// Force conversion if both types is numeric but not equal.
+        if (in_elem.type->behavesAsNumber() && out_elem.type->behavesAsNumber() && !out_elem.type->equals(*in_elem.type))
```

Comment:
Note, that in task about INSERT SELECT, we have requirement to convert only to larger numeric type.
Nevertheless, current implementation is Ok.